### ISSUE TITLE
Simpler 'Dupable' typeclass

### ISSUE THREAD

**Issue by utdemir** [tweag/linear-base#151]
Currently `Dupable` typeclass requires implementing the `dupV` function:

```haskell
class Consumable a => Dupable a where
  dupV :: KnownNat n => a #-> V n a
```

However, to me, it's not obvious how that can be implemented; and after spending hours on it today I couldn't figure it out. However

```haskell
dup2 :: Dupable a => a #-> (a, a)
```

function is simpler to implement, and I think we should define `Dupable` in terms of this. `Dupable`'s laws are already defined in terms of `dup2` instead of `dupV`, which also points to `dup2` being the simpler option.

I had this issue when trying to implement a `Dupable` instance for `Data.Array.Mutable.Linear`. The existing `Dupable` implementations all have simple implementations implemented using `V`'s applicative instance (or using other `Dupable` instances), but we can not use that if we can not share the value. In `Array`'s case, it's easy to implement a function like `clone :: Array a #-> (Array a, Array a)`, but going from that to `dupV` was not easy (for me).

We should still keep `dupV` in the class and define `{-# MINIMAL dup2 | dupV #-}`, in case when `dupV` is more efficient.

**Comment by aspiwack** [comment]
There are a lot of design questions here, to be honest.

First, the vector-of-known-length library, while very useful in principle with linear types is… not great. It won't be super efficient, and the API is not the best (some stuff is missing, that's for sure). It's fine to include `dup2` into the typeclass for a quick-and-dirty definition.

By the way, one of the reason for moving away from `dup2` is that `V n` is a (data) applicative functor, and that really helps defining instances (of course, we could use `V 2` everywhere, instead, that would work too, but, in a sense, you may want more than 2, and contrary to monoids, you can't really rely on lists to represent your output, and generalise from `dup2` nicely, hence the `V n`).

However, there is a hitch in this story: `Data.Applicative` are kind of meh for strict structure. Basically, a data applicative is a container `c` equipped with `zipWithN :: (a_1 #-> … a_n #-> b) -> c a_1 #-> … c a_N #-> c b` for every `N`. But the implementation of `zipWithN` in terms of the binary `(<*>)` creates a ton of intermediate structure. It's alright when the structure is lazy, and built one data constructor at a time: there is a bit of superfluous allocation, but at least it's very short lived. But for a vector, you have to build N intermediate vectors. And it's rather unfortunate. If we were to write `zipWithN` manually, we don't need these intermediate structures. But there are infinitely many, so we may not want to actually add them all to the type class.

The vector library, I think, behaves reasonably well when chaining `zip`-s, because of vector fusion (a powerful, if unreliable, ally). And I can't help but thinking: where there is array fusion, there is polarised “fusion”. Could there be a variant of `(<*>)` for vectors of known length (or `zip` for arrays, I suppose), which chains well, using polarised arrays instead.

And down in the rabbit hole in 3 hops. Not bad, don't you think|

**Comment by utdemir** [comment]
Thank you for the detailed explanation (and I agree that it is a decent rabbit hole).

> First, the vector-of-known-length library, while very useful in principle with linear types is… not great. It won't be super efficient, and the API is not the best (some stuff is missing, that's for sure). It's fine to include dup2 into the typeclass for a quick-and-dirty definition.

Improving the API of the vector-of-known-length library would definitely be useful, and solve this issue. As I said, the reason I created this issue is just because I couldn't implement `dupV` using the current API (even after exporting a few internal utilities like `caseNat`). However, I think that would require a much longer discussion and careful thought; so adding `dup2` inside the typeclass might be an easy improvement until that discussion.

> By the way, one of the reason for moving away from dup2 is that V n is a (data) applicative functor, and that really helps defining instances.

I agree here. The discussion on the original change (https://github.com/tweag/linear-base/pull/26) also helps. I think we definitely should keep `dupV` inside the typeclass too.

> It's alright when the structure is lazy, and built one data constructor at a time: there is a bit of superfluous allocation, but at least it's very short lived. But for a vector, you have to build N intermediate vectors.
> ... Could there be a variant of (<*>) for vectors of known length (or zip for arrays, I suppose), which chains well, using polarised arrays instead.

I agree. It is a bit out of scope for this issue, but I was initally thinking about using regular lists there, especially since the current implementation uses `cons` a lot. But using polarized arrays might be better, especially if we reimplement a more efficient version of the current constructor which doesn't allocate a new vector for each element.

**Comment by Divesh-Otwani** [comment]
> By the way, one of the reason for moving away from `dup2` is that `V n` is a (data) applicative functor, and that really helps defining instances (of course, we could use `V 2` everywhere, instead, that would work too, but, in a sense, you may want more than 2, and contrary to monoids, you can't really rely on lists to represent your output, and generalise from `dup2` nicely, hence the `V n`).

How important is this? `Consumable` and `Dupable` are fundamental to linear programming. If we requires new users to be familiar with some fancy type level stuff to merely use `Dupable`, this puts a large barrier to entry. 

If we don't already have a reasonable design for doing a n-dimensional zip on data applicatives, then changing `Dupable` for that doesn't seem prudent to me.

I'd like to _strongly_ push for just using `dup2` until we have something substantial with (1) changing the known-length vector interface AND (2) the motivating story of how this helps define instances with n-dimensional zips for data applicatives.

At least for the release, I think we should table the fancier version. Granted, this is a strong opinion -- nonetheless, I think we need to move experimental and perhaps awesome things to branches and keep simple and robust stuff in master. Thoughts, disagreements, agreements, concerns, manifestos, vendettas? @aspiwack @utdemir

**Comment by facundominguez** [comment]
I'm formally linking the discussion on linear `Foldable/Traversal` [here](https://github.com/tweag/linear-base/pull/126#issuecomment-686508060).

> By the way, one of the reason for moving away from dup2 is that V n is a (data) applicative functor, and that really helps defining instances

It looks like this would be true for Arrays if `traversable` worked with data functors, for instance.

That said, I sympathize with the approach suggested by @Divesh-Otwani. IIUC, `V n` has potential to avoid creating intermediate copies but this needs to be better motivated, and the interface needs to be easy to use.

**Comment by aspiwack** [comment]
Ok, let me explain in a bit more details where this all comes from. Because I've obviously not made a great job of it.

But first, let me preface that I realise that it may not be possible to have a satisfactory interface for `V n`, as it pushes against the limits of GHC's typechecker. Still there is a reason why I went there.

I'm also not agreeing with having `dup2` part of the type class. @utdemir convinced me that it was useful, if only because it gives an easy way to define custom instances. So if we are to persist with `dupV`, we absolutely want the type class to be `{-# MINIMAL dup2|dupV #-}`.

## The motivation

The design of `Dupable` around vectors of known length comes from two angle: defining instances and using `dup*` functions.

### Creating instances

Basically, we want (well, I want, to be honest) instances at composite types to be convenient to write.

Compare defining `dup2` and `dupV` for `(a, b, c)`

```haskell
dup2 (x, y, z) =
  dup2 x & \case (x1, x2) ->
  dup2 y & \case (y1, y2) ->
  dup2 z & \case (z1, z2) ->
  ((x1, y1, z1), (x1, y2, z2))
  
dupV (x, y, z) = (,,) <$> dupV x <*> dupV y <*> dupV z
```

I think it's uncontroversial to say that the latter is quite a bit preferable (to be fair, originally, we hadn't come up with the `& \case` trick, so it was even worse, the case is not as strong today, but still rather strong).

Of course, we can solve this by introducing a simple type

```haskell
newtype V2 a = V2 (a, a)
instance Applicative V2
```

We don't need any abstraction on natural numbers just for the applicative syntax

### The asymmetry between monoids and comonoids

We are used to defining monoids in terms of binary operations. That's because when we need to multiply more than two elements, the syntax is actually fairly elegant: `x⋆y⋆z⋆w`. In Haskell, we also often rely on list with `mconcat`.

Not so for comonoids. We would have to do things like

```haskell
dup4 =
  let
    !(x, x') = u
    !(y, y') = x'
    !(z, w) = y'
  in
  (x, y, z, w)
```

(this is the dual of `x⋆y⋆z⋆w`). We can't rely on lists either, because

```haskell
 let ![x, y, z, w] = dupL u
```

is an incomplete pattern.

So the Right Abstraction™ is to have an n-ary (co)operation. And have the output be a vector of known length.

I think that this conclusion is unescapable. Until someone finds a composable way too chain binary (or something else) comonoid operations, which is easier to use within Haskell. But if I was the betting type, I'd put some money on it not happening.

## The difficulties with vectors of known length

There is a first difficulty that I mentioned above: the applicative instance for `dupV` is not actually very efficient. But I've gone back to primary sources since, and I now realise that it is sufficient to have a pull-array version of vectors of known length: they have an efficient applicative instance (presumably, we would then replace `dupV` by `dupP :: KnownNat n => a #-> Pull.V n a` as a method of the instance).

So I think this is fine.

I don't think that there is much difficulties for the end user either: I expect that 99% of the uses of `dupV` (or `dupP`), will be either through the applicative instance, or via a concrete number literal (_e.g._ `dupV @4`). Neither requiring anything that GHC is not great at.

On the other hand, writing the library for vectors of known length can be a bit of a tangle. Because at least some of these functions will require GHC to reason beyond its capabilities. Which we can solve either by using a plugin or by using `unsafeCoerce`.

The limitation of using plugins is that Linear Haskell is work in progress, so there will always be a version of linear base depending on a recent master. Where plugins have a tendency to be broken. So, unless we roll our own plugin within linear base (I don't think it's a good idea), we would end up with a linear-base which doesn't compile.

So it's got to be `unsafeCoerce`. Which is always unpleasant, and may even make things unpleasantly inefficient (it is allowed to prevent inlining for instance). For instance, I'm not very happy with the implementation of the elimination function, currently. But maybe we can build confidence that we can actually make such functions efficient. And live with the handful of `unsafeCoerce` that would be needed for the full featured library.

**Comment by facundominguez** [comment]
View patterns make the expression somewhat better
```Haskell
dup4 = \case
    (dup2 -> (x, dup2 -> (y, dup2 -> (z, w)))) -> (x, y, z, w)
```
If the main motivation to use fixed-length vectors is a syntax problem, it looks like too much of a hassle. But it would be fine nonetheless to go with `{-# MINIMAL dup2|dupV #-}` if someone is *that* motivated to work on the `dupV` thing.

**Comment by aspiwack** [comment]
I want to acknowledge that view patterns are indeed a corresponding syntactic construct to chaining application.

They don't work with linear types yet, but that's sort of besides the point.

Still, I'm not quite ready to abandon the vectors of known length there. I've become quite convinced in the past few weeks that vectors of known length are easy to use as long as you use concrete length, which is how I intend them to be used (except for `dupV`, but here as well, we intend to never use type-level arithmetic). It's writing generic library function for them which can be a little challenging. But that's on us.

Besides, vectors of known length are the natural zippy thing in a linear types context since we can't have a total `ZipList` data applicative. See also #208 .

**Comment by Divesh-Otwani** [comment]
I have decided to reopen this issue because I'm still a bit confused:

- We've established the need for `dupV` is not dependent on elegant syntax since we can give `V2` an applicative instance that's suitably efficient
- We've established that because of how co-monoids work, we'd want `dupV` because writing `dup4` with just `dup2` (and so on) is both messy and inefficient.

My question is why do we want `dup4` to begin with? It intuitively seems to me that all use cases pretty much either need a `Movable` instance or just need `dup2`. 

Can we provide a practical use of `dup4`? @aspiwack @utdemir
